atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -434,6 +437,96 @@ public void markTableInvolved(TableReference tableRef) {
         return postFilteredResults;
     }
 
+    @Override
+    public Iterator<Map.Entry<Cell, byte[]>> getSortedColumns(
+            TableReference tableRef, Iterable<byte[]> rows, BatchColumnRangeSelection columnRangeSelection) {
+        checkGetPreconditions(tableRef);
+        if (Iterables.isEmpty(rows)) {
+            return Collections.emptyIterator();
+        }
+        Iterable<byte[]> distinctRows = getDistinctRows(rows);
+
+        hasReads = true;
+        int batchSize = getPerRowBatchSize(columnRangeSelection, Iterables.size(distinctRows));
+        BatchColumnRangeSelection perBatchSelection = BatchColumnRangeSelection.create(
+                columnRangeSelection.getStartCol(), columnRangeSelection.getEndCol(), batchSize);
+
+        Map<byte[], RowColumnRangeIterator> rawResults =
+                keyValueService.getRowsColumnRange(tableRef, distinctRows, perBatchSelection, getStartTimestamp());
+
+        return getPostFilteredSortedColumns(tableRef, columnRangeSelection, distinctRows, rawResults);
+    }
+
+    private Iterator<Map.Entry<Cell, byte[]>> getPostFilteredSortedColumns(
+            TableReference tableRef,
+            BatchColumnRangeSelection columnRangeSelection,
+            Iterable<byte[]> distinctRows,
+            Map<byte[], RowColumnRangeIterator> rawResults) {
+        Comparator<Cell> cellComparator = columnOrderThenPreserveInputRowOrder(distinctRows);
+
+        Iterator<Map.Entry<Cell, Value>> postFilterIterator = getRowColumnRangePostFilteredWithoutSorting(
+                tableRef, mergeColumnFirst(rawResults.values(), cellComparator), columnRangeSelection.getBatchHint());
+        Iterator<Map.Entry<Cell, byte[]>> remoteWrites = Iterators.transform(
+                postFilterIterator,
+                entry -> Maps.immutableEntry(entry.getKey(), entry.getValue().getContents()));
+        Iterator<Map.Entry<Cell, byte[]>> localWrites =
+                getSortedColumnsLocalWrites(tableRef, distinctRows, columnRangeSelection, cellComparator);
+        Iterator<Map.Entry<Cell, byte[]>> merged = IteratorUtils.mergeIterators(
+                localWrites,
+                remoteWrites,
+                Comparator.comparing(Map.Entry::getKey, cellComparator),
+                com.palantir.util.Pair::getLhSide);
```

Comment:
This is a fairly common pattern (it's used twice more in this class: once in getRowsColumnRange and once in mergeInLocalWrites). You'll probably need to allow the user to pass the comparator in, but the logical piece of "when you need to merge, local writes win" is something that can be shared across the iterators.